hs_labels/models.py

```diff
@@ -0,0 +1,876 @@
+        """
+        if not isinstance(this_label, basestring):
+            raise HSLUsageException("Label is not text")
+
+        label_string = UserLabels.clean_label(this_label)                 # remove leading and trailing spaces
+
+        return BaseResource.objects.filter(rlabels__rl2url__ulabels=self,
+                                           rlabels__rl2url__kind=LabelCodes.LABEL,
+                                           rlabels__rl2url__label__exact=label_string)\
+                                        .distinct()\
+                                        .order_by('rlabels__rl2url__label')
+    @property
+    def user_labels(self):
+        return UserResourceLabels.objects.values_list('label', flat=True)\
+                                 .filter(ulabels=self, kind=LabelCodes.LABEL)\
+                                 .distinct().order_by('label')
+
+    ######################################
+    # Label a resource
+    ######################################
+    @staticmethod
+    def clean_label(name):
+        label_string = re.sub('/', r'', name)                   # no /'s
+        label_string = label_string.strip()                     # no leading or trailing whitespace
+        label_string = re.sub(r'\s+', r' ', label_string)       # collapse multiple whitespace, including tabs
+        return label_string
+
+    def label_resource(self, this_resource, this_label):
+        """
+        Share a resource with a group
+
+        :param this_resource: Resource to share.
+        :param this_label: Label to assign to resource
+        :return: None
+
+        Users are allowed to label any resource, including resources to which they do not have access.
+        This is not an access control problem because labeling information is private.
+        """
+
+        # check for user error
+        if not isinstance(this_resource, BaseResource):
+            raise HSLUsageException("Target is not a resource")
+        label_resource = this_resource.rlabels
+
+        if not isinstance(this_label, basestring):
+            raise HSLUsageException("Label is not text")
+
+        # remove leading and trailing spaces
+        label_string = UserLabels.clean_label(this_label)
+
+        # This logic implicitly limits one to one record per resource and requester.
+        try:
+            UserResourceLabels.objects.get(rlabels=label_resource,
+                                           label__exact=label_string,
+                                           ulabels=self)
+            # if this succeeds, resource is already labeled with this label_string.
+
+        # only create label if it does not exist. No duplicates.
+        except UserResourceLabels.DoesNotExist:
+            # create a new label record
+            UserResourceLabels(rlabels=label_resource,
+                               kind=LabelCodes.LABEL,
+                               label=label_string,
+                               ulabels=self).save()
```

Comment:
`UserResourceLabels.objects.create()` would make it easier to read this as creating a new instance.

Also maybe you could use `get_or_create()` above to automatically create it instead?